### PR TITLE
Add Python multiline strings

### PR DESCRIPTION
**Pull request by thalesmello**
This pull request detects multiline strings in python
- Tries to guess whether the text is SQL and set filetype accordingly
- Otherwise will not assign any filetype

An option to control SQL detection is also included

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -59,6 +59,32 @@ If you like the plugin, consider rating it on [vim.org](http://www.vim.org/scrip
...
+  proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
+
+  Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor
+  incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis
+  nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
+  Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu
+  fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
+  culpa qui officia deserunt mollit anim id est laborum.
+  """)
+  ```
```
Comment (AndrewRadev) [inline README.markdown:86]: Could we have a shorter example for this print? Maybe we could have just one lorem block with a paragraph break in the middle, around "Duis"?

File: autoload/inline_edit.vim [hunk 2]
```diff
@@ -172,3 +172,84 @@ function! inline_edit#HereDoc()
 
   return [start, end, filetype, indent]
 endfunction
+
+" function! inline_edit#PythonQuotedString() {{{2
+
+function s:check_inside_python_string()
+  return index(map(synstack(line('.'), col('.')), 'synIDattr(v:val, "name")'), "pythonString") >= 0
+endfunction
```
Comment (AndrewRadev) [inline autoload/inline_edit.vim:180]: Usually, I keep script-local/private functions after the public ones. Could you move this to the bottom of the file? It would also be nice if you could change the under_scored naming style to a CamelCased function.

File: autoload/inline_edit.vim [hunk 3]
```diff
@@ -172,3 +172,84 @@ function! inline_edit#HereDoc()
...
+" github-flavoured markdown.
+function! inline_edit#PythonMultilineString()
+  call inline_edit#PushCursor()
+
+  normal! 0
+
+  if !s:check_inside_python_string()
+    call inline_edit#PopCursor()
+    return []
+  endif
```
Comment (AndrewRadev) [inline autoload/inline_edit.vim:191]: One way to avoid popping the cursor before every return is to do it in a try-finally block:

``` vim
call inline_edit#PushCursor()

try
  "do stuff
  return []
finally
  call inline_edit#PopCursor()
endtry
```

In this case, you can wrap the entire contents of the function in the try-finally and pop the cursor at the end, since you only use it to maintain the same position, and don't modify it.

File: autoload/inline_edit.vim [hunk 4]
```diff
@@ -172,3 +172,84 @@ function! inline_edit#HereDoc()
...
+
+  let end = search(end_quote, 'W')
+
+  if end == 0
+    " No end quote was found
+    call inline_edit#PopCursor()
+    return []
+  endif
+
+  let end = end - 1
```
Comment (AndrewRadev) [inline autoload/inline_edit.vim:228]: You can shorten this to `let end -= 1`

File: autoload/inline_edit.vim [hunk 5]
```diff
@@ -172,3 +172,84 @@ function! inline_edit#HereDoc()
...
+  let start = line('.') + 1
+
+  if quote_type == 1
+    let end_quote = '"""'
+  elseif quote_type == 2
+    let end_quote = "'''"
+  else
+    echoerr "Invalid quote pattern found"
+    return []
+  endif
```
Comment (AndrewRadev) [inline autoload/inline_edit.vim:218]: This code confuses me, because the `quote_type` variable is the result of the `search()` call and that should be the line number of the match, and it doesn't seem to be used like this here. Could you explain how the `quote_type` variable takes the value of 1 or 2 depending on the match? Maybe I'm missing something in the variable assignments.

File: autoload/inline_edit.vim (hunk 3 above)
Comment (thalesmello) [reply autoload/inline_edit.vim:191]: That's cool! Never thought of using finally blocks like that in VimScript

File: autoload/inline_edit.vim (hunk 5 above)
Comment (thalesmello) [reply autoload/inline_edit.vim:218]: According to `:help search()`, when the. `p` flag is used, the return value will be the something equivalent to the index of the match group inside the pattern. More specifically: `1` if the pattern matches but no subgroup matches, `2` if the first (and only) subgroup matched.

I use that to specifically identify whether `"""` (return value of `1`) or `'''` (return value of `2`) is being used in the multiline string. Afterwards it will try to find the matching triple multiline string mark.

Notice I didn't use `searchpair()` here because the `start` and `end` patterns are the same, and the function doesn't work in that case, as it assumes the end pattern to actually be a nested pair.

**Comment by thalesmello** [comment]
@AndrewRadev I do agree with changing the SQL filetype heuristic to be case sensitive, as that's the pattern most people use in the industry and will reduce the change of false positives.

I've pushed all the changes you asked for. Let me know what you think.

File: autoload/inline_edit.vim (hunk 5 above)
Comment (AndrewRadev) [reply autoload/inline_edit.vim]: Ah, I see -- I didn't know about the "p" flag or had forgotten about it, it makes a lot more sense now :+1:.
